ENH: linalg: Add LAPACK wrappers for trexc and trsen.

Sbte opened the pull request:
#### Reference issue
Refs #10881

#### What does this implement/fix?
Adds wrappers for trexc and trsen which are required for sorting Schur decompositions. Wrappers for the generalized counterparts already exist, which these implementations are based upon.

swallan commented:
Something to consider for added robustness is including the examples that the NAG manual has in the test suite:

- [dtrexc- real](https://www.nag.com/numeric/nl/nagdoc_27/clhtml/f08/f08qfc.html) 
- [ztrexc - complex](https://www.nag.com/numeric/nl/nagdoc_27/clhtml/f08/f08qtc.html)  
- [dtrsen - real](https://www.nag.com/numeric/nl/nagdoc_27/clhtml/f08/f08qgc.html)
- [ztrsen - complex](https://www.nag.com/numeric/nl/nagdoc_27/clhtml/f08/f08quc.html)

Sbte commented:
I added the requested tests. I also added those tests for tgexc and tgsen, but these functions do not give unique answers (1 * 1 == -1 * -1), and hence the tests failed on my system. Therefore I did not submit them.

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -42,15 +42,54 @@ subroutine <prefix2>gejsv(joba,jobu,jobv,jobr,jobt,jobp,m,n,a,lda,sva,u,ldu,v,ld
+    ! Reorder the Schur decomposition of a real matrix
+    ! using an orthogonal or unitary equivalence transformation.
+
+    callstatement { ifst++; ilst++; (*f2py_func)((wantq?"V":"N"),&n,a,&lda,q,&ldq,&ifst,&ilst,work,&info); }
```
We try to stay away from this `++` business as it often makes it more difficult instead of easy if you want to take one input or output from one routine to another. Better leave it as is and deal with it elsewhere.

Sbte replied:
Is elsewhere in the code of the user? So if they want to swap row 1 and 2 (in 0 based indices) they have to pass 2 and 3 instead?

ilayn replied:
Yes pretty much if you work with LAPACK routine you comply with Fortran stuff. So that strange conversions don't get in the way.

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -42,15 +42,54 @@ subroutine <prefix2>gejsv(joba,jobu,jobv,jobr,jobt,jobp,m,n,a,lda,sva,u,ldu,v,ld
+    integer optional,intent(in),check(wantq==0||wantq==1) :: wantq=1
+    integer intent(hide),depend(a) :: n=shape(a,1)
+    <ftype2> intent(in,out,copy),dimension(lda,n) :: a
+    integer intent(hide),depend(a) :: lda=shape(a,0)
```
all workspace vars need to be at least 1 so `MAX(1, shape(a,0))` same for ldq below

Sbte replied:
In that case, shouldn't it be a `check()` instead?

ilayn replied:
`a` can be empty but `lda` cannot be 0 even when `a` is empty.

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -42,15 +42,54 @@ subroutine <prefix2>gejsv(joba,jobu,jobv,jobr,jobt,jobp,m,n,a,lda,sva,u,ldu,v,ld
+    integer intent(hide),depend(q) :: ldq=shape(q,0)
+    integer intent(in) :: ifst
+    integer intent(in) :: ilst
+    <ftype2> intent(hide),dimension(n) :: work
```
add `depend(n)`

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -42,15 +42,54 @@ subroutine <prefix2>gejsv(joba,jobu,jobv,jobr,jobt,jobp,m,n,a,lda,sva,u,ldu,v,ld
+    ! Reorder the Schur decomposition of a complex matrix
+    ! using an orthogonal or unitary equivalence transformation.
+
+    callstatement { ifst++; ilst++; (*f2py_func)((wantq?"V":"N"),&n,a,&lda,q,&ldq,&ifst,&ilst,&info); }
```
same ++ business here too

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -42,15 +42,54 @@ subroutine <prefix2>gejsv(joba,jobu,jobv,jobr,jobt,jobp,m,n,a,lda,sva,u,ldu,v,ld
+
+    integer optional,intent(in),check(wantq==0||wantq==1) :: wantq=1
+    integer intent(hide),depend(a) :: n=shape(a,1)
+    <ftype2c> intent(in,out,copy),dimension(lda,n) :: a
```
same lda and ldq issue here

ilayn reviewed (changes requested):
Few things about the trexc parts but trsen looks OKish. I'll try to sweep again on Wednesday.

Sbte commented:
Thanks for the additional clarifications. I made the suggested changes and also modified tgexc accordingly.

Sbte commented:
@ilayn Anything else I can do to get this merged? I want to make a release of my software and need these functions to be available from scipy.

ilayn commented:
My sincere apologies for the ridiculous delay. Lately, I haven't been up to my obligations as a maintanier.

Let's get first started with rebasing onto main branch from your old master (we renamed our running branch from `master` to `main`). And we'll take it from there as quickly as we can. 

If you encounter difficulties just open a new PR to save time and effort. You don't need to wrestle with git anyways.

Sbte commented:
@ilayn thanks for looking into this. I rebased the patches and all checks have succeeded.

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -42,27 +42,66 @@ subroutine <prefix2>gejsv(joba,jobu,jobv,jobr,jobt,jobp,m,n,a,lda,sva,u,ldu,v,ld
     ! Reorder the generalized Schur decomposition of a real matrix
     ! pair using an orthogonal or unitary equivalence transformation.
 
-    callstatement { ifst++; ilst++; (*f2py_func)(&wantq,&wantz,&n,a,&lda,b,&ldb,q,&ldq,z,&ldz,&ifst,&ilst,work,&lwork,&info); }
```
This unfortunately we cannot do immediately because it breaks backward compatibility. I also think it should have been done as such but historical reasons.

Sbte replied:
I added all of those changes after your suggestions. Back when I added them they didn't break backwards compatibility, because I added the original code (that is changed here) in the same unreleased version. But now there have indeed been some releases in between. If you want I can change the new wrappers to be the same as the old ones. That was what I also did in the original pull request before modifying them like this according to your comments. I think the wrappers should at least be consistent with each other.

ilayn replied:
Then I am talking nonsense and don't put enough attention. I'll get back to this asap.

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -72,26 +111,138 @@ subroutine <prefix2c>tgexc(wantq,wantz,n,a,lda,b,ldb,q,ldq,z,ldz,ifst,ilst,info)
     ! Reorder the generalized Schur decomposition of a complex matrix
     ! pair using an orthogonal or unitary equivalence transformation.
 
-    callstatement { ifst++; ilst++; (*f2py_func)(&wantq,&wantz,&n,a,&lda,b,&ldb,q,&ldq,z,&ldz,&ifst,&ilst,&info); }
```
Same here

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -42,27 +42,66 @@ subroutine <prefix2>gejsv(joba,jobu,jobv,jobr,jobt,jobp,m,n,a,lda,sva,u,ldu,v,ld
 
-    integer intent(hide),check(wantq==0||wantq==1) :: wantq=1
-    integer intent(hide),check(wantz==0||wantz==1) :: wantz=1
+    integer optional,intent(in),check(wantq==0||wantq==1) :: wantq=1
```
This also needs some care because now we are changing the signature of the wrapper.

Sbte replied:
I think this is fine because the argument is optional and was previously hidden. Any old call should still work.

ilayn reviewed (changes requested):
I'll check the trsen part in a bit weekend is a bit occupied.

ilayn commented on scipy/linalg/flapack_other.pyf.src:
```diff
@@ -72,26 +111,138 @@ subroutine <prefix2c>tgexc(wantq,wantz,n,a,lda,b,ldb,q,ldq,z,ldz,ifst,ilst,info)
+
+    character optional,intent(in),check(*job=='N'||*job=='E'||*job=='V'||*job=='B'):: job = 'B'
+    logical intent(in),dimension(n),depend(n) :: select
+    <ftype2> intent(in),dimension(n,n) :: a
```
For `lwork` queries, the only things needed are `job`, `compq` and `n`. That avoids having a huge matrix passed to f2py and possibly converted only to be discarded later. This is safe because the unneeded arguments are not referenced and we find those variables by checking the Fortran code directly. 

We do this trick in many places (or at least in the ones we touched in the last years). Here is one example where everything else is dummified to scalars and `work` is shortened https://github.com/scipy/scipy/blob/main/scipy/linalg/flapack_other.pyf.src#L508=

Sbte replied:
I just copy-pasted it from `tgsen_lwork` that you implemented a few years ago: https://github.com/scipy/scipy/commit/b6414d9accda02ab30a3d38b994c5224ede4bd43. I think `trsen` and `tgsen` should use the same calling conventions. Should I also modify `tgsen_lwork`?

ilayn replied:
Yes but in trsen and tgsen the array itself is referenced so we have to pass them. These are not one-size-fits-for-all wrappers unfortunately. LAPACK people do strange things with each routine but probably they have a good reason. The remaining ones shouldn't need those.

Sbte replied:
Sorry, I don't really understand. I only implement an `_lwork` method for `trsen`. `trsen` doesn't seem to reference `a` when `lwork=-1`, but `tgsen` does, because it has to compute `m`. I could make `trsen_lwork` to not require `a`, but then you'd get something like

```
    lwork = _compute_lwork(tgsen_lwork, select, s, t)
```
and
```
    lwork = _compute_lwork(trsen_lwork, select, t.shape[0])
```
instead of
```
    lwork = _compute_lwork(trsen_lwork, select, t)
```
which seems inconsistent.

ilayn replied:
There is no consistency required for these wrappers because this is inconsistent in the LAPACK anyways. If you have a C-contiguous array from numpy and you want to compute the lwork you have to pass the whole array and f2py will happily create another copy of this array then Fortran routine won't even touch it. So we hide as much as we can. No consistency is sought here but mostly convenience and performance. Hence instead of the whole array we use `n` which is in my opinion much simpler.

Sbte replied:
Actually, this won't work. DTRSEN also uses `t`:

https://netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_gafe383d8eb4c3815d8946a5e24ba1899a.html

So I think this is fine as is.

ilayn replied:
Trust me it does not 

```fortran

       wantbh = lsame( job, 'B' )
       wants = lsame( job, 'E' ) .OR. wantbh
       wantsp = lsame( job, 'V' ) .OR. wantbh
       wantq = lsame( compq, 'V' )
 *
       info = 0
       lquery = ( lwork.EQ.-1 )
       IF( .NOT.lsame( job, 'N' ) .AND. .NOT.wants .AND. .NOT.wantsp )
      $     THEN
          info = -1
       ELSE IF( .NOT.lsame( compq, 'N' ) .AND. .NOT.wantq ) THEN
          info = -2
       ELSE IF( n.LT.0 ) THEN
          info = -4
       ELSE IF( ldt.LT.max( 1, n ) ) THEN
          info = -6
       ELSE IF( ldq.LT.1 .OR. ( wantq .AND. ldq.LT.n ) ) THEN
          info = -8
       ELSE
 *
 *        Set M to the dimension of the specified invariant subspace,
 *        and test LWORK and LIWORK.
 *
          m = 0
          pair = .false.
          DO 10 k = 1, n
             IF( pair ) THEN
                pair = .false.
             ELSE
                IF( k.LT.n ) THEN
                   IF( t( k+1, k ).EQ.zero ) THEN
                      IF( SELECT( k ) )
      $                  m = m + 1
                   ELSE
                      pair = .true.
                      IF( SELECT( k ) .OR. SELECT( k+1 ) )
      $                  m = m + 2
                   END IF
                ELSE
                   IF( SELECT( n ) )
      $               m = m + 1
                END IF
             END IF
    10    CONTINUE
 *
          n1 = m
          n2 = n - m
          nn = n1*n2
 *
          IF( wantsp ) THEN
             lwmin = max( 1, 2*nn )
             liwmin = max( 1, nn )
          ELSE IF( lsame( job, 'N' ) ) THEN
             lwmin = max( 1, n )
             liwmin = 1
          ELSE IF( lsame job, 'E' ) ) THEN
             lwmin = max( 1, nn )
             liwmin = 1
          END IF
 *
          IF( lwork.LT.lwmin .AND. .NOT.lquery ) THEN
             info = -15
          ELSE IF( liwork.LT.liwmin .AND. .NOT.lquery ) THEN
             info = -17
          END IF
       END IF
 *
       IF( info.EQ.0 ) THEN
          work( 1 ) = lwmin
          iwork( 1 ) = liwmin
       END IF
 *
       IF( info.NE.0 ) THEN
          CALL xerbla( 'DTRSEN', -info )
          RETURN
       ELSE IF( lquery ) THEN
          RETURN
       END IF
```

ilayn replied:
wait why are we changing the variable names ? I was under the impression that a was not referenced but this is a and q now for some reason. Let's not do that please.

Sbte replied:
I don't see your comment here, on github, but did receive it in the mail. `t` is referenced here in the code you copy-pasted:

```
 *
 *        Set M to the dimension of the specified invariant subspace,
 *        and test LWORK and LIWORK.
 *
          m = 0
          pair = .false.
          DO 10 k = 1, n
             IF( pair ) THEN
                pair = .false.
             ELSE
                IF( k.LT.n ) THEN
                   IF( t( k+1, k ).EQ.zero ) THEN
```

and the tests indeed start failing when I don't pass `t`.

ilayn replied:
Yes because I was looking for `a` but somehow it is renamed see my comment above

Sbte replied:
Ah, yes, that's my bad. I did that because of tgsen, but LAPACK names them differently (a and b for tgsen, t for trsen). I'll change it to be t, thanks!

ilayn reviewed (changes requested):
I think most of this is ready to go except one issue with how we add lwork routines.

ev-br commented:
As a general question, would it make sense at some point to start wrapping cython_lapack implementations instead of f2py ones?

ilayn commented:
I would say yes if we can do it peacefully but then what do we do with the existing ones is my counter-question. Because so far these served OK however we are indeed coming to a point where historical artifacts and new ones are getting mixed up slowly. An algorithmic wrapper-automation in a different namespace might be an option. 

I used to think that hiding most of these internal variables was a really cool feature but come to think of it we are the odd ones out and every other fortran wrapper project is going for a full signature exposure and letting the user fill in those pesky details.  Also without f2py support mechanism the user has to pick the right dtype and contiguousness manually for the wrappers to work. Now all is taken care of by these wrappers.

ev-br commented:
> but then what do we do with the existing ones is my counter-question. 

Nothing, absolutely nothing. They exist, they work, they are not broken, hence they are in no need of doing anything about. 

All I'm saying maybe after this PR is merged, for the next lapack python wrapper we try to wrap the cython interface. It'll need all sorts checks f2py is doing, so maybe it'll turn out to be a futile exercise. But the only way is to give it a go and see I guess. 

Anyway, it all is OT for this pull request, I'll stop :-).

ilayn commented:
OK fingers crossed. Let's keep an eye on this in case we missed anything before 1.10.x goes out. 

Thank you @Sbte and once again sorry for the delay it took to come back to this.

Sbte commented:
No problem! Glad it's in now!
